File: apps/web/src/lib/analytics.ts
import { getToken } from "@/lib/auth";
import { getApiBase } from "@/lib/config";

export type RoiKpi = {
  monthly_savings_rub: number;
  monthly_savings_delta_pct: number | null;
  vehicle_count: number;
  active_vehicle_count: number;
  active_routes_count: number;
  orders_today: number;
  fuel_alerts_count: number;
  fuel_alerts_critical: number;
  roi_percent_annual: number | null;
  payback_months: number | null;
  annual_savings_rub: number;
  subscription_cost_monthly_rub: number;
  breakdown: {
    fuel: number;
    route: number;
    dispatch: number;
    maintenance: number;
  };
};

export async function fetchRoiKpi(): Promise<RoiKpi> {
  const token = getToken();
  if (!token) throw new Error("Not authenticated");

  const response = await fetch(`${getApiBase()}/api/v1/analytics/roi`, {
    headers: { Authorization: `Bearer ${token}` },
  });

  if (!response.ok) {
    throw new Error("Failed to load KPI data");
  }

  return response.json();
}

export function formatRub(value: number): string {
  return new Intl.NumberFormat("ru-RU", {
    style: "currency",
    currency: "RUB",
    maximumFractionDigits: 0,
  }).format(value);
}

File: apps/api/app/schemas/analytics.py
from pydantic import BaseModel, Field


class SavingsBreakdown(BaseModel):
    fuel: float = 0
    route: float = 0
    dispatch: float = 0
    maintenance: float = 0
    permit: float = 0


class KpiDataSource(BaseModel):
    vehicles: str = "database"
    monthly_savings: str = "database"
    fuel_alerts: str = "database"
    active_routes: str = "database"
    orders_today: str = "database"
    roi: str = "calculated"


class RoiKpiResponse(BaseModel):
    monthly_savings_rub: float
    monthly_savings_delta_pct: float | None = None
    vehicle_count: int
    active_vehicle_count: int
    active_routes_count: int
    orders_today: int
    fuel_alerts_count: int
    fuel_alerts_critical: int
    roi_percent_annual: float | None = None
    payback_months: float | None = None
    annual_savings_rub: float
    subscription_cost_monthly_rub: float = Field(default=49_000)
    breakdown: SavingsBreakdown
    data_sources: KpiDataSource
